keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,401 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Enabling pagination typically improves performance by reducing response sizes.
+However, the kube-apiserver's caching mechanism (watchcache) does not currently
+support pagination. When a client requests a paginated list, the limit parameter
+might be ignored, or the request might bypass the cache and be served directly
+from etcd, leading to performance issues. This document proposes an enhancement
+to the kube-apiserver's caching layer to enable efficient serving of paginated
+lists from the cache, improving performance and predictability.
+
+## Motivation
+
+The kube-apiserver's watchcache is an eventually consistent cache of the cluster
+state. It allows the API server to serve read requests without constantly hitting
+etcd. However, watchcache is not designed for pagination.
+
+Pagination allows clients to request large lists of resources in smaller chunks.
+The client specifies a limit (maximum number of objects per page) and receives
+a continue token in the response. This token is then used in subsequent requests
+to fetch the next page.
+
+To serve paginated requests correctly, the API server needs access to a consistent
+snapshot of the data from which the initial page was generated. Watchcache does
+not store historical data. Therefore, currently, the API server does the following:
+
+- For `resourceVersion=""`: Serve paginated requests directly from etcd, bypassing the cache.
+- For `resourceVersion="0"`: Ignore the limit parameter and return the entire list in a single response.
+
+Both options are problematic:
+
+- Serving from etcd negates the benefits of the cache,
+  leading to increased load on both the API server and etcd.
+- Ignoring the limit parameter defeats the purpose of pagination,
+  potentially returning massive responses that can overwhelm clients.
+
+This situation leads to unexpected behavior for users,
+where enabling pagination can degrade performance instead of improving it.
+
+### Goals
+
+- Improve the performance of paginated list requests to be comparable to non-paginated lists.
+- Make pagination behavior predictable and consistent.
+- Reduce load on etcd caused by paginated requests.
+
+### Non-Goals
+
+- Serve `resourceVersion="N"` request from watch cache
+- Support indices when paginating.
+- Eliminate all paginated list requests to etcd.
+
+## Proposal
+
+Leveraging the recent rewrite of the watchcache storage layer to use a B-tree
+(https://github.com/kubernetes/kubernetes/pull/126754), we propose to utilize
+B-tree snapshots to serve paginated lists.
+
+Mechanism:
+1. **Snapshot Creation:** When a paginated list request (with a limit parameter
```

Comment:
In a HA configuration where there are multiple apiservers and client requests are load balanced across those apiservers, is the idea that each apiserver creates the snapshot on the first paginated request it receives, even if the request is for a subsequent page?

Reply:
Not really - if I get only a subsequent request for n-th page, I simply forward it to etcd....

... unless we go with the approach I'm proposing instead: https://github.com/kubernetes/enhancements/pull/5017#discussion_r1909953609